apollo-router/src/query_planner/bridge_query_planner.rs

```diff
@@ -47,28 +48,55 @@ pub(crate) struct BridgeQueryPlanner {
 
 impl BridgeQueryPlanner {
     pub(crate) async fn new(
-        schema: String,
+        sdl: String,
         configuration: Arc<Configuration>,
     ) -> Result<Self, ServiceBuildError> {
-        let planner = Arc::new(
-            Planner::new(
-                schema.clone(),
-                QueryPlannerConfig {
-                    incremental_delivery: Some(IncrementalDeliverySupport {
-                        enable_defer: Some(configuration.supergraph.defer_support),
-                    }),
-                },
-            )
-            .await?,
-        );
+        let schema = Schema::parse(&sdl, &configuration)?;
+
+        let planner = Planner::new(
+            sdl,
+            QueryPlannerConfig {
+                incremental_delivery: Some(IncrementalDeliverySupport {
+                    enable_defer: Some(configuration.supergraph.defer_support),
+                }),
+                graphql_validation: matches!(
+                    configuration.experimental_graphql_validation,
+                    GraphQLValidation::Legacy | GraphQLValidation::Both
+                ),
+            },
+        )
+        .await;
+
+        let planner = match planner {
+            Ok(planner) => planner,
+            Err(err) => {
+                if configuration.experimental_graphql_validation == GraphQLValidation::Both {
+                    let has_validation_errors = err.iter().any(|err| err.is_validation_error());
+
+                    if has_validation_errors != schema.has_errors() {
+                        tracing::warn!(
+                            monotonic_counter.apollo_router_validation_false_negative = 1,
+                            "query planner reported a validation error, but apollo-rs did not"
+                        );
+                    }
+                }
+
+                return Err(err.into());
+            }
+        };
+
+        if configuration.experimental_graphql_validation == GraphQLValidation::Both {
+            tracing::warn!(
+                monotonic_counter.apollo_router_validation_false_positive = 1,
+                "query planner did not report validation error, but apollo-rs did"
```

Comment:
Doesn’t this need a `schema.has_errors()` condition?

Reply:
Oh yeah. this was an `assert!(!schema.has_errors())` first and I lost it when changing to a trace!